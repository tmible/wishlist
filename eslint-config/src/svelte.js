import svelte from "eslint-plugin-svelte";
import globals from "globals";

export default [
  ...svelte.configs["flat/recommended"],
  {
    languageOptions: {
      globals: {
        ...globals.browser,
        ...globals.node,
      },
      ecmaVersion: "latest",
      sourceType: "module",
      parserOptions: {
        extraFileExtensions: [ ".svelte" ],
      },
    },

    settings: {
      "import/parsers": {
        "svelte-eslint-parser": [ ".svelte" ],
        "espree": [ ".js" ],
      },
    },

    rules: {
      /* import */
      // svelte does tree shaking when compiling
      "import/no-namespace": "off",

      /* svelte */
      "svelte/infinite-reactive-loop": "error",
      "svelte/no-dom-manipulating": "error",
      "svelte/no-dupe-on-directives": "error",
      "svelte/no-dupe-use-directives": "error",
      "svelte/no-reactive-reassign": "error",
      "svelte/no-store-async": "error",
      "svelte/require-store-callbacks-use-set-param": "error",
      "svelte/block-lang": "error",
      "svelte/no-ignored-unsubscribe": "error",
      "svelte/no-immutable-reactive-statements": "error",
      "svelte/no-inline-styles": "error",
      "svelte/no-reactive-functions": "error",
      "svelte/no-reactive-literals": "error",
      "svelte/no-useless-mustaches": "error",
      "svelte/require-each-key": "error",
      "svelte/require-event-dispatcher-types": "error",
      "svelte/require-stores-init": "error",
      "svelte/valid-each-key": "error",
      "svelte/derived-has-same-inputs-outputs": "error",
      "svelte/first-attribute-linebreak": "error",
      "svelte/html-closing-bracket-new-line": "error",
      "svelte/html-closing-bracket-spacing": "error",
      "svelte/html-quotes": "error",
      "svelte/html-self-closing": [
        "error",
        {
          "void": "never",
          "normal": "never",
          "svg": "always",
          "math": "never",
          "component": "always",
          "svelte": "always",
        },
      ],
      "svelte/indent": [ "error", { ignoredNodes: [ "ConditionalExpression" ] } ],
      "svelte/max-attributes-per-line": [ "error", { multiline: 1, singleline: 100 } ],
      "svelte/mustache-spacing": "error",
      "svelte/no-extra-reactive-curlies": "error",
      "svelte/no-spaces-around-equal-signs-in-attribute": "error",
      "svelte/prefer-class-directive": "error",
      "svelte/prefer-style-directive": "error",
      "svelte/shorthand-attribute": "error",
      "svelte/sort-attributes": [
        "error",
        {
          order: [
            // `this` property.
            "this",
            // `bind:this` directive.
            "bind:this",
            // `id` attribute.
            "id",
            // `name` attribute.
            "name",
            // `slot` attribute.
            "slot",
            // `--style-props`
            { match: "/^--/u", sort: "ignore" },
            // `style` attribute, and `style:` directives.
            [ "style", "/^style:/u" ],
            // `class` attribute.
            "class",
            // `class:` directives.
            { match: "/^class:/u", sort: "ignore" },
            // other attributes.
            {
              match: [ "!/:/u", "!/^(?:this|id|name|style|class)$/u", "!/^--/u" ],
              sort: "ignore",
            },
            // `bind:` directives (other then `bind:this`), and `on:` directives.
            [ "/^bind:/u", "!bind:this", "/^on:/u" ],
            // `use:` directives.
            { match: "/^use:/u", sort: "ignore" },
            // `transition:` directive.
            { match: "/^transition:/u", sort: "ignore" },
            // `in:` directive.
            { match: "/^in:/u", sort: "ignore" },
            // `out:` directive.
            { match: "/^out:/u", sort: "ignore" },
            // `animate:` directive.
            { match: "/^animate:/u", sort: "ignore" },
            // `let:` directives.
            { match: "/^let:/u", sort: "ignore" },
          ],
        },
      ],
      "svelte/spaced-html-comment": "error",
      "svelte/no-inner-declarations": "error",
      "svelte/no-trailing-spaces": "error",
    },
  },
  {
    files: [ "**/*.svelte" ],
    rules: {
      // svelte components can import themselves
      "import/no-self-import": "off",
    },
  },
  {
    files: [ "**/**.test.js", "**/**.test.svelte.js", "**/__tests__/**" ],
    settings: { "import/ignore": [ "vitest", "@testing-library/svelte" ] },
  },
];
