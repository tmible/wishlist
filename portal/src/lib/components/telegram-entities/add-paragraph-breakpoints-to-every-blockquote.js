/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

// eslint-disable-next-line no-secrets/no-secrets -- Просто длинное название
/**
 * Обеспечение наличия разметки абзацев в каждой цитате. Начала в начале и конца в конце.
 * Функция для использования в {@link Array.prototype.flatMap}
 * @function addParagraphBreakpointsToEveryBlockquote
 * @param {Breakpoint} breakpoint Точка изменения разметки в тексте
 * @param {unknown} _ Не используется
 * @param {Breakpoint[]} breakpoints Все точки изменения разметки в тексте
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
export const addParagraphBreakpointsToEveryBlockquote = (breakpoint, _, breakpoints) => [
  breakpoint,
  ...(
    breakpoint.entity.type === 'blockquote' &&
    !breakpoints.some(
      ({ offset, entity, type }) => (
        breakpoint.offset === offset &&
        entity.type === 'paragraph' &&
        breakpoint.type === type
      ),
    ) ?
      [{
        offset: breakpoint.offset,
        entity: {
          type: 'paragraph',
          length: breakpoint.length,
        },
        type: breakpoint.type,
      }] :
      []
  ),
];
