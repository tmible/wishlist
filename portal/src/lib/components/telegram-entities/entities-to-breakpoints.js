/** @typedef {import('./parser.svelte').Entity} Entity */
/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Отображение элементов разметки в объекты, представляющие точки изменения разметки в тексте
 * @function entitiesToBreakpoints
 * @param {Entity[]} entities Элементы разметки
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
export const entitiesToBreakpoints = (entities) => entities.flatMap((entity) => [
  { offset: entity.offset, entity, type: 'opening' },
  { offset: entity.offset + entity.length, entity, type: 'closing' },
]);
