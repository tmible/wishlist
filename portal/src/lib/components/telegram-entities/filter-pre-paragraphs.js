/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Обеспечение отсутствия разметки абзацев в каждом блоке кода. Начала в начале и конца в конце.
 * Функция для использования в {@link Array.prototype.filter}
 * @function filterPreParagraphs
 * @param {Breakpoint} breakpoint Точка изменения разметки в тексте
 * @param {unknown} _ Не используется
 * @param {Breakpoint[]} breakpoints Все точки изменения разметки в тексте
 * @returns {boolean} Признак удовлетворения элемента фильтру
 */
export const filterPreParagraphs = (breakpoint, _, breakpoints) => (
  breakpoint.entity.type !== 'paragraph' ||
    !breakpoints.some(({ offset, entity, type }) => (
      breakpoint.offset === offset &&
        entity.type === 'pre' &&
        breakpoint.type === type
    ))
);
