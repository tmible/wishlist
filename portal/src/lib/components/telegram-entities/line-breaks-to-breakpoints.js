/** @typedef {ReturnType<typeof RegExp.prototype.exec>} Match */
/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Отображение переводов строк в объекты, представляющие точки
 * изменения разметки в тексте для каждого перевода строки
 * @function flatMapLineBreaks
 * @param {Match[]} lineBreaks Объекты нахождения переводов строк с помощью регулярного выражения
 * @param {number} textLength Длина всего текста
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
const flatMapLineBreaks = (lineBreaks, textLength) => lineBreaks.flatMap(({ index }, i) => [{
  offset: index,
  entity: {
    type: 'paragraph',
    length: index - (i > 0 ? lineBreaks[i - 1].index + 1 : 0),
  },
  type: 'closing',
}, {
  offset: index,
  entity: {
    type: 'paragraph',
    length: 1,
  },
  type: 'opening',
}, {
  offset: index + 1,
  entity: {
    type: 'paragraph',
    length: 1,
  },
  type: 'closing',
}, {
  offset: index + 1,
  entity: {
    type: 'paragraph',
    length: (
      i < lineBreaks.length - 1 ? lineBreaks[i + 1].index + 1 : textLength
    ) - (index + 1),
  },
  type: 'opening',
}]);

/**
 * Отображение переводов строк в объекты, представляющие точки изменения
 * разметки в тексте. В частности начала и концы абзацев
 * @function lineBreaksToBreakpoints
 * @param {Match[]} lineBreaks Объекты нахождения переводов строк с помощью регулярного выражения
 * @param {number} textLength Длина всего текста
 * @returns {Breakpoint[]} Точки изменения разметки в тексте
 */
export const lineBreaksToBreakpoints = (lineBreaks, textLength) => [
  ...(
    textLength > 0 ?
      [{
        offset: 0,
        entity: {
          type: 'paragraph',
          length: lineBreaks.length > 0 ? lineBreaks[0].index + 1 : textLength,
        },
        type: 'opening',
      }] :
      []
  ),
  ...flatMapLineBreaks(lineBreaks, textLength),
  ...(
    textLength > 0 ?
      [{
        offset: textLength,
        entity: {
          type: 'paragraph',
          length: textLength - (lineBreaks.length > 0 ? lineBreaks.at(-1).index + 1 : 0),
        },
        type: 'closing',
      }] :
      []
  ),
];
