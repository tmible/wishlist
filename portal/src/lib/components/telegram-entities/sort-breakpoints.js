/** @typedef {import('./parser.svelte').Breakpoint} Breakpoint */

/**
 * Компаратор точек изменения разметки в тексте по положению в тексте (по возрастанию)
 * @function compareByOffset
 * @param {Breakpoint} a Первая сравниваемая точка
 * @param {Breakpoint} b Вторая сравниваемая точка
 * @returns {number | null} Отрицательное число, если a раньше b; положительное, если b раньше a;
 * null, если отступы от начала текста равны
 */
const compareByOffset = (a, b) => (a.offset === b.offset ? null : a.offset - b.offset);

/**
 * Компаратор точек изменения разметки в тексте по типу. Сначала закрывающие, потом открывающие
 * @function compareByType
 * @param {Breakpoint} a Первая сравниваемая точка
 * @param {Breakpoint} b Вторая сравниваемая точка
 * @returns {-1 | 1 | null} -1, если a раньше b; 1, если b раньше a; null, если типы равны
 */
const compareByType = (a, b) => (a.type === b.type ? null : (a.type === 'closing' ? -1 : 1));

/**
 * Компаратор точек изменения разметки в тексте по типу соответсвующей сущности. Приоретизация
 * цитат, блоков кода в первую очередь и абзацев во вторую. Открывающие в прямом порядке
 * приоритетов, закрывающие -- в обратном
 * @function filterPreBlockquotesAndParagraphs
 * @param {Breakpoint} a Первая сравниваемая точка
 * @param {Breakpoint} b Вторая сравниваемая точка
 * @returns {-1 | 1 | null} -1, если a раньше b; 1, если b раньше a; null, если типы равны
 */
const filterPreBlockquotesAndParagraphs = (a, b) => {
  if ([ 'pre', 'blockquote' ].includes(a.entity.type)) {
    return a.type === 'closing' ? 1 : -1;
  }
  if ([ 'pre', 'blockquote' ].includes(b.entity.type)) {
    return b.type === 'closing' ? -1 : 1;
  }
  if (a.entity.type === 'paragraph') {
    return a.type === 'closing' ? 1 : -1;
  }
  if (b.entity.type === 'paragraph') {
    return b.type === 'closing' ? -1 : 1;
  }
  return null;
};

/**
 * Компаратор точек изменения разметки в тексте по длине соответсвующей сущности.
 * Открывающие в порядке убывания длины, закрывающие -- в порядке возрастания
 * @function orderByEntityLength
 * @param {Breakpoint} a Первая сравниваемая точка
 * @param {Breakpoint} b Вторая сравниваемая точка
 * @returns {number} Отрицательное число, если a раньше b; положительное, если b раньше a;
 * 0, если длины равны
 */
const orderByEntityLength = (a, b) => (
  a.type === 'closing' ?
    a.entity.length - b.entity.length :
    b.entity.length - a.entity.length
);

/**
 * Компаратор точек изменения разметки в тексте [по положению в тексте]{@link compareByOffset},
 * [по типу]{@link compareByType},
 * [по типу соответсвующей сущности]{@link filterPreBlockquotesAndParagraphs},
 * [по длине соответсвующей сущности]{@link orderByEntityLength}
 * @function sortBreakpoints
 * @param {Breakpoint} a Первая сравниваемая точка
 * @param {Breakpoint} b Вторая сравниваемая точка
 * @returns {number} Отрицательное число, если a раньше b; положительное, если b раньше a;
 * 0 иначе
 */
export const sortBreakpoints = (a, b) => (
  compareByOffset(a, b) ??
  compareByType(a, b) ??
  filterPreBlockquotesAndParagraphs(a, b) ??
  orderByEntityLength(a, b)
);
