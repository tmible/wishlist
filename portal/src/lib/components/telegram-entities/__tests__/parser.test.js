// @vitest-environment jsdom
import { cleanup, render } from '@testing-library/svelte';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { addParagraphBreakpointsToEveryBlockquote } from '../add-paragraph-breakpoints-to-every-blockquote.js';
import { entitiesToBreakpoints } from '../entities-to-breakpoints.js';
import { filterPreParagraphs } from '../filter-pre-paragraphs.js';
import { lineBreaksToBreakpoints } from '../line-breaks-to-breakpoints.js';
import TelegramEntitiesParser from '../parser.svelte';
import { sortBreakpoints } from '../sort-breakpoints.js';

vi.mock(
  '../add-paragraph-breakpoints-to-every-blockquote.js',
  async (importOriginal) => {
    const original = await importOriginal();
    const spy = vi.fn(original.addParagraphBreakpointsToEveryBlockquote);
    return { addParagraphBreakpointsToEveryBlockquote: spy };
  },
);
vi.mock(
  '../entities-to-breakpoints.js',
  async (importOriginal) => {
    const original = await importOriginal();
    const spy = vi.fn(original.entitiesToBreakpoints);
    return { entitiesToBreakpoints: spy };
  },
);
vi.mock(
  '../filter-pre-paragraphs.js',
  async (importOriginal) => {
    const original = await importOriginal();
    const spy = vi.fn(original.filterPreParagraphs);
    return { filterPreParagraphs: spy };
  },
);
vi.mock(
  '../line-breaks-to-breakpoints.js',
  async (importOriginal) => {
    const original = await importOriginal();
    const spy = vi.fn(original.lineBreaksToBreakpoints);
    return { lineBreaksToBreakpoints: spy };
  },
);
vi.mock(
  '../sort-breakpoints.js',
  async (importOriginal) => {
    const original = await importOriginal();
    const spy = vi.fn(original.sortBreakpoints);
    return { sortBreakpoints: spy };
  },
);

/* eslint-disable-next-line @stylistic/js/max-len --
   Длинная константа, читать не нужно, хранить так проще */
const entities = JSON.parse('[{"type":"blockquote","offset":105,"length":33},{"type":"bold","offset":0,"length":7},{"type":"bold","offset":49,"length":21},{"type":"bot_command","offset":159,"length":5},{"type":"cashtag","offset":154,"length":4},{"type":"code","offset":34,"length":6},{"type":"email","offset":165,"length":17},{"type":"hashtag","offset":147,"length":6},{"type":"italic","offset":8,"length":6},{"type":"italic","offset":71,"length":13},{"type":"mention","offset":138,"length":8},{"type":"phone_number","offset":195,"length":12},{"type":"pre","offset":92,"length":13,"language":null},{"type":"spoiler","offset":41,"length":7},{"type":"spoiler","offset":117,"length":14},{"type":"strikethrough","offset":24,"length":9},{"type":"strikethrough","offset":56,"length":7},{"type":"strikethrough","offset":119,"length":5},{"type":"text_link","offset":208,"length":8,"url":"https://example.com"},{"type":"underline","offset":15,"length":8},{"type":"underline","offset":78,"length":14},{"type":"url","offset":183,"length":11}]');
const text =
  `ueyrghi erogij woeriugh woerifrjh woeifj woeihjf woeifn woerifj woefij woiefj owiejf weiufn
oreigjnpoerkglsrgnj
eporkg erkmg wpoerg wpoef
@wlkergn #erkbn $KDF /ernm ev@jewpeorij.lnjf example.com +12345678901 ldtkrhnj`;

describe('telegram entities parser', () => {
  afterEach(() => {
    vi.clearAllMocks();
    cleanup();
  });

  it('should convert entities to breakpoints', () => {
    render(TelegramEntitiesParser, { entities, text });
    expect(entitiesToBreakpoints).toHaveBeenCalledWith(entities);
  });

  it('should consider line breaks as breakpoints', () => {
    render(TelegramEntitiesParser, { entities, text });
    expect(
      lineBreaksToBreakpoints,
    ).toHaveBeenCalledWith(
      Array.from(text.matchAll(/(?!^)\n(?!$)/g)),
      text.length,
    );
  });

  it('should add paragraph breakpoints to every blockquote', () => {
    render(TelegramEntitiesParser, { entities, text });
    expect(addParagraphBreakpointsToEveryBlockquote).toHaveBeenCalled();
  });

  it('should filter pre paragraphs', () => {
    render(TelegramEntitiesParser, { entities, text });
    expect(filterPreParagraphs).toHaveBeenCalled();
  });

  it('should sort breakpoints', () => {
    render(TelegramEntitiesParser, { entities, text });
    expect(sortBreakpoints).toHaveBeenCalled();
  });

  it('should build and display tree', () => {
    const { container } = render(TelegramEntitiesParser, { entities, text });
    expect(container.innerHTML).toMatchSnapshot();
  });
});
