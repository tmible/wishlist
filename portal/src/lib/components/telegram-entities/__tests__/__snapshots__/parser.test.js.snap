// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`telegram entities parser > should build and display tree 1`] = `"<!----><!----><!----><p class="break-words"><!----><!----><!----><b><!----><!---->ueyrghi</b><!----> <!----><!----><i><!----><!---->erogij</i><!----> <!----><!----><u><!----><!---->woeriugh</u><!----> <!----><!----><s><!----><!---->woerifrjh</s><!----> <!----><!----><code><!----><!---->woeifj</code><!----> <!----><!----><span class="text-transparent hover:text-base-content spoiler" data-entity="spoiler"><!----><!---->woeihjf</span><!----> <!----><!----><b><!----><!---->woeifn <!----><!----><s><!----><!---->woerifj</s><!----> woefij</b><!----> <!----><!----><i><!----><!---->woiefj <!----><!----><u><!----><!---->owiejf</u></i><!----><!----><u><!----><!----> weiufn</u></p><!----><!----><pre><!----><!---->oreigjnpoerkg</pre><!----><!----><blockquote class="prose"><!----><!----><!----><p class="break-words"><!----><!---->lsrgnj</p><!----><!----><p class="break-words"><!----><!---->epork<!----><!----><span class="text-transparent hover:text-base-content spoiler" data-entity="spoiler"><!----><!---->g <!----><!----><s><!----><!---->erkmg</s><!----> wpoerg</span><!----> wpoef</p></blockquote><!----><!----><p class="break-words"><!----><!----><!----><a href="https://t.me/wlkergn" data-entity="mention"><!----><!---->@wlkergn</a><!----> <!----><!----><span class="link cursor-auto" data-entity="hashtag"><!----><!---->#erkbn</span><!----> <!----><!----><span class="link cursor-auto" data-entity="cashtag"><!----><!---->$KDF</span><!----> <!----><!----><span class="link cursor-auto" data-entity="bot_command"><!----><!---->/ernm</span><!----> <!----><!----><a href="mailto:ev@jewpeorij.lnjf" data-entity="email"><!----><!---->ev@jewpeorij.lnjf</a><!----> <!----><!----><a href="https://example.com"><!----><!---->example.com</a><!----> <!----><!----><a href="tel:+12345678901" data-entity="phone_number"><!----><!---->+12345678901</a><!----> <!----><!----><a href="https://example.com" data-entity="text_link"><!----><!---->ldtkrhnj</a></p>"`;
